testing: set fake user name and email in .gitconfig

stefano-garzarella opened the pull request:
If the system git user name and email are not set, the create_git_repo() fails with this error:
```
> ./testing/run_tests.sh

*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: unable to auto-detect email address
````

This PR set a fake git user name and email in the fake home directory to solve this issue and add a little change to testing/0000-gitconfig-home to work with this change.

stefano-garzarella commented:
forced push just to change the title of the first patch of this PR

stefanha commented:
Is this still needed after:

commit 47f66c23fc8d09ac8955e279bf07a305123f5eef
Author: Eduardo Habkost <ehabkost@redhat.com>
Date:   Tue Oct 8 18:30:57 2019 -0300

    Test cases: override $HOME so 'git-config' works

?

stefano-garzarella commented:
> Is this still needed after:
> 
> commit [47f66c2](https://github.com/stefanha/git-publish/commit/47f66c23fc8d09ac8955e279bf07a305123f5eef)
> Author: Eduardo Habkost [ehabkost@redhat.com](mailto:ehabkost@redhat.com)
> Date: Tue Oct 8 18:30:57 2019 -0300
> 
> ```
> Test cases: override $HOME so 'git-config' works
> ```
> 
> ?

I think it's this commit that creates the problem in my system, where the user name and email are set via git config --global (~/.gitconfig).

However, I think it is correct to create a new temporary HOME, but in my case, this generates the problem described.

As an alternative, we could set the user name and email in the new repository via git config --local, but if new tests create new repositories, we should do it every time.

stefanha commented:
On Mon, Nov 4, 2019, 09:15 Stefano Garzarella <notifications@github.com>
wrote:

> Is this still needed after:
>
> commit 47f66c2
> <https://github.com/stefanha/git-publish/commit/47f66c23fc8d09ac8955e279bf07a305123f5eef>
> Author: Eduardo Habkost ehabkost@redhat.com
> Date: Tue Oct 8 18:30:57 2019 -0300
>
> Test cases: override $HOME so 'git-config' works
>
> ?
>
> I think it's this commit that creates the problem in my system, where the
> user name and email are set via git config --global (~/.gitconfig).
>
> However, I think it is correct to create a new temporary HOME, but in my
> case, this generates the problem described.
>
> As an alternative, we could set the user name and email in the new
> repository via git config --local, but if new tests create new
> repositories, we should do it every time.
>

Once HOME is set, the name and email can be configured with --global. Then
there is no need to mv the real ~/.gitconfig temporarily.

HOME should be exported for the duration of the test so that all test code
and subprocesses use the temporary .gitconfig.

Does this sound like a good fix?

>

stefano-garzarella commented:
On Mon, Nov 4, 2019 at 9:22 AM Stefan Hajnoczi <notifications@github.com> wrote:
>
> On Mon, Nov 4, 2019, 09:15 Stefano Garzarella <notifications@github.com>
> wrote:
>
> > Is this still needed after:
> >
> > commit 47f66c2
> > <https://github.com/stefanha/git-publish/commit/47f66c23fc8d09ac8955e279bf07a305123f5eef>
> > Author: Eduardo Habkost ehabkost@redhat.com
> > Date: Tue Oct 8 18:30:57 2019 -0300
> >
> > Test cases: override $HOME so 'git-config' works
> >
> > ?
> >
> > I think it's this commit that creates the problem in my system, where the
> > user name and email are set via git config --global (~/.gitconfig).
> >
> > However, I think it is correct to create a new temporary HOME, but in my
> > case, this generates the problem described.
> >
> > As an alternative, we could set the user name and email in the new
> > repository via git config --local, but if new tests create new
> > repositories, we should do it every time.
> >
>
> Once HOME is set, the name and email can be configured with --global. Then
> there is no need to mv the real ~/.gitconfig temporarily.
>
> HOME should be exported for the duration of the test so that all test code
> and subprocesses use the temporary .gitconfig.
>
> Does this sound like a good fix?


I don't follow you... I think this is the problem of GitHub PR vs
e-mail based review or I wrote the wrong commit message.

I'll try to summarize; this PR contains 2 patches:
1. testing/gitconfig: backup .gitconfig before modify it
   This patch backups the fake .gitconfig, because the test expects to
find only the "gitpublishtesting.readingworks=yes".
   Since I'm adding also the user name and email, this test failed.
2. testing: set fake user name and email in .gitconfig
   This patch does just what you said, setting user name and email
with --global, because we currently export the temp HOME to all test
code and subprocesses in run_test.sh.

Maybe I should change the patch 1, and instead of backup the fake
.gitconfig, I should check if it contains
"gitpublishtesting.readingworks=yes" plus other stuff.

What do you suggest?

ehabkost commented:
> Maybe I should change the patch 1, and instead of backup the fake .gitconfig, I should check if it contains "gitpublishtesting.readingworks=yes" plus other stuff.

This sounds better to me.

ehabkost commented on testing/run_tests.sh:
```diff
@@ -92,6 +92,11 @@ export HOME="$RESULTS_DIR/home"
+
+# set fake user name and email to make git happy if system values are not set
+git config --global user.email "git-publish@tes.ts"
+git config --global user.name "git-publish tests"
```
I feel nervous seeing `git config` run before the `0000-gitconfig-home` test case runs and confirms the `$HOME` override is really working as expected and not touching the real home directory. What about writing to `$RESULTS_DIR/home` directly, just in case? e.g.:

```bash
cat >> "$RESULTS_DIR/home/.gitconfig" <<EOF
[user]
email = git-publish-test@example.com
name = git-publish tests
EOF
```

stefano-garzarella replied:
I felt the same, I think it's better as you suggested.
Thanks, I'll fix this!

stefano-garzarella commented:
> Just replacing `git config --global --show-origin --list` with `git config --global --show-origin --list | grep readingworks` would probably be enough, and simpler than forcibly backing up and restoring the existing config file.

Thanks for the review, yes this was the idea :-)

stefano-garzarella commented:
Force-pushed a new version following the @ehabkost 's suggestions:
- Patch 1: replaced `git config --global --show-origin --list` with `git config --global --show-origin --list | grep readingworks` 
- Patch 2: replaced `git config --global user...` to set user name and email, writing directly in the `"$RESULTS_DIR/home/.gitconfig" `

stefanha commented:
Thanks @stefano-garzarella and @ehabkost
